### PR TITLE
Organize cout and logs

### PR DESCRIPTION
**Pull request by fabiencastan**
connected to poparteu/openMVG#97

### CODE REVIEW

File: src/openMVG/prettyprint.hpp [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+//
+// A pretty printing library for C++
+//
+// Usage:
+// Include this header, and operator<< will "just work".
+
+#ifndef H_PRETTY_PRINT
```
Comment (simogasp) [inline src/openMVG/prettyprint.hpp]: `#pragma once`?

File: .travis.yml [hunk 2]
```diff
@@ -36,6 +36,11 @@ env:
     - BUILD_PROCESSOR="`uname -p`"
     - OPENMVG_SOURCE=${TRAVIS_BUILD_DIR}/src
     - OPENMVG_BUILD=${TRAVIS_BUILD_DIR}/build
+    # CMAKE
+    - CMAKE_URL="https://cmake.org/files/v3.4/cmake-3.4.1-Linux-x86_64.tar.gz"
```
Comment (simogasp) [inline .travis.yml]: https://cmake.org/files/v3.0/cmake-3.0.0.tar.gz in case we want to use the minimum version

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -625,7 +626,7 @@ MESSAGE("** Use OpenCV SIFT features: " ${OpenMVG_USE_OCVSIFT})
 MESSAGE("** Use CCTAG markers: " ${OpenMVG_USE_CCTAG})
 MESSAGE("** Build Alembic exporter: " ${OpenMVG_USE_ALEMBIC})
 MESSAGE("** Use OpenGV for rig localization: " ${OpenMVG_USE_OPENGV})
-MESSAGE("** Use POPART logger: " ${OpenMVG_USE_POPART_LOG})
+MESSAGE("** Use POPART logger: " ${OpenMVG_USE_OPENMVG_LOG})
```
Comment (simogasp) [inline src/CMakeLists.txt]: also remove POPART

File: .travis.yml [hunk 4]
```diff
@@ -78,6 +83,17 @@ before_install:
...
+ - >
+    if [ "$(ls -A ${CMAKE_INSTALL})" ]; then
+      echo "CMake found in cache.";
+    else
+      mkdir --parent ${CMAKE_SOURCE}
+      mkdir --parent ${CMAKE_INSTALL}
+      travis_retry wget --no-check-certificate --quiet -O - ${CMAKE_URL} | tar --strip-components=1 -xz -C ${CMAKE_INSTALL}
+    fi
+ - export PATH=${CMAKE_INSTALL}/bin:${PATH};
+ - cmake --version
```
Comment (simogasp) [inline .travis.yml]: at https://github.com/poparteu/openMVG/blob/new_enhanced_logs/.travis.yml#L191 remember to add `log` as module to build for boost

File: src/openMVG/logger.hpp [hunk 5]
```diff
@@ -1,14 +1,47 @@
...
-//
-#ifdef WANTS_POPART_COUT
-#define POPART_COUT(x) std::cout << x << std::endl
-#define POPART_CERR(x) std::cerr << x << std::endl
-#define POPART_COUT_DEBUG(x) std::cout << x << std::endl
+#define OPENMVG_COUT(x) std::cout << x << std::endl
+#define OPENMVG_CERR(x) std::cerr << x << std::endl
+
+#ifdef WANTS_OPENMVG_COUT
+#ifdef HAVE_BOOST
```
Comment (simogasp) [inline src/openMVG/logger.hpp]: maybe indent the #if #else to improve readability

File: .travis.yml (hunk 2 above)
Comment (fabiencastan) [inline .travis.yml]: I will use 3.4 as it's the minimal version required for popsift and there is a WIP PR to have popsift integrated.

Comment (simogasp) [inline .travis.yml]: ok, but it won't be a problem to use a different one, as the PopsiftConfig generate something compatible with >= 2.8.12

Comment (fabiencastan) [inline .travis.yml]: Yes, but I would like to build popsift on the fly in travis.

**Comment by simogasp** [comment]
@fabiencastan 
we could add in env
```
- OPENMVG_INSTALL=${OPENMVG_BUILD}/install
```
and then in `before_script:`
```
 cmake \
       -DCMAKE_INSTALL_PREFIX= $OPENMVG_INSTALL \
       -DCMAKE_BUILD_TYPE=$BUILD_TYPE \
       ...
```
and do
```
make install -j 2 VERBOSE=1
```
so that we can check the installation too.

**Comment by simogasp** [comment]
@zvrba We were discussing with @fabiencastan about the organization of the logging. I was suggesting something in this spirit in order to be able to support other logging system (ie glog, glog is interesting because ceres already use it, so we will able to get the logs from that too).

in Cmake

```cmake
OPTION(OpenMVG_USE_LOG "Enable logger" ON)

# a variable to set the type of logger to use
set(OpenMVG_LOGGER "LOG_COUT" CACHE STRING "The logger to use for openMVG"))
set(LOGGERS_Values "LOG_COUT;LOG_BOOST;LOG_GLOG" CACHE INTERNAL "List of possible values for the OpenMVG_LOGGER cache variable”)
set_property(CACHE OpenMVG_LOGGER PROPERTY STRINGS ${LOGGERS_Values})
...

IF(OpenMVG_USE_BOOST)
  if(OpenMVG_LOGGER STREQUAL "LOG_BOOST")
    FIND_PACKAGE(Boost 1.53.0 QUIET COMPONENTS system filesystem program_options thread serialization log log_setup)
  else()
    FIND_PACKAGE(Boost 1.53.0 QUIET COMPONENTS system filesystem program_options thread serialization)
  endif()
  #maybe we can switch to REQUIRED, at least the user see that something is wrong...
  IF(Boost_FOUND)
    .. 
  else()
    #fallback to cout
    set(OpenMVG_LOGGER "cout")
    ..
  endif()
endif()

#the same for GLog
if(OpenMVG_LOGGER STREQUAL "LOG_GLOG")
  find_package(Glog REQUIRED)
  ...
endif()

IF(OpenMVG_USE_LOG)
  ADD_DEFINITIONS(-DWANTS_OPENMVG_COUT)
  ADD_DEFINITIONS(-D${OpenMVG_LOGGER})
ENDIF()
```

then inside logger.h

```c++
#ifdef WANTS_OPENMVG_COUT
  #ifdef LOG_BOOST
    #define BOOST_LOG_DYN_LINK 1
    #include <boost/log/trivial.hpp>

    #define OPENMVG_LOG_TRACE_OBJ BOOST_LOG_TRIVIAL(trace)
    ...
    
  #elif defined(LOG_GLOG)
    #include <glog.hpp>
    #define OPENMVG_LOG_TRACE_OBJ GLOG_LOG_TRIVIAL(trace)
    ...
  #else //cout
    #define OPENMVG_LOG_TRACE_OBJ std::cout
    ...
  #endif
  #define OPENMVG_LOG(MODE, ...) MODE << __VA_ARGS__
#else
  #define OPENMVG_LOG_TRACE_OBJ std::stringstream
  ...
#endif
```

What do you think? Too complicated? :-)

**Comment by zvrba** [comment]
No, it's not too complicated, this is the right way to do it. Minor nitpick: the last one (disabled logging) has to be `#define OPENMVG_LOG_TRACE_OBJ std::stringstream()` (with trailing parentheses, so that it actually creates a temporary object).

Now, several things:

I think that Boost should be a mandatory dependency for openMVG. It's a free, easily obtainable library with unproblematic license and there's no need to complicate openMVG build and code by making it optional.

Logging to `std::cout` should not be an option at all. It is easy to configure boost.log to redirect its output to arbitrary location (including syslog, windows event log, etc.) plus it supports powerful run-time filtering w/minimal configuration which doesn't require recompilation. (http://www.boost.org/doc/libs/1_62_0/libs/log/doc/html/log/tutorial/trivial_filtering.html ; `log_setup` has a utility function which compiles simple string expressions to filter objects.)

I'm unsure whether to offer the choice between glog and boost.log. Standardizing on ONE logging library would be the best, if at all possible.

I have a strong bias towards boost.log; I've used it in other popart projects as well as HAL code, and I'm really happy with it. I'd be most happy if glog could be replaced with boost.log in Ceres. In the worst case, configure Ceres' logging so that it just forwards log messages to boost.log, if it's not too much work.

**Comment by zvrba** [comment]
While we're at boost: I'll think about whether it's possible to detect whether boost is compiled with c++11 during configuration phase, so that the user doesn't have to set `BOOST_NO_CXX11` flag manually.

**Comment by simogasp** [comment]
> No, it's not too complicated, this is the right way to do it. Minor nitpick: the last one (disabled logging) has to be #define OPENMVG_LOG_TRACE_OBJ std::stringstream() (with trailing parentheses, so that it actually creates a temporary object).

actually it doesn't really matter what you put there since ` OPENMVG_LOG` is defined as empty, those OBJs need to contain something only to assure that the OPENMVG_LOG_*() make sense (i guess @fabiencastan )

> I think that Boost should be a mandatory dependency for openMVG. It's a free, easily obtainable library with unproblematic license and there's no need to complicate openMVG build and code by making it optional.

yes it can actually replace many small dependencies we have. WIP

> Logging to std::cout should not be an option at all. It is easy to....

that was my idea, i thought it would be nice to have a fallback on cout just in case. I've just sketched here how to deal with multiple solutions as a way to organize things but honestly, as you said, I don't know if down the road things will become nasty and managing 3 different things will become more a problem than a feature. The glog thing was really because of Ceres, nothing else. FInding a way to redirect the output on boost would be great.

**Pull request merged**